toolchain/base/value_store.h

```diff
@@ -171,62 +168,60 @@ class ValueStore
   // for (auto [id, value] : store.enumerate()) { ... }
   // ```
   auto enumerate() const -> auto {
-    UnpoisonAll();
     auto index_to_id = [](auto pair) -> std::pair<IdT, ConstRefType> {
       auto [index, value] = pair;
-      return std::pair<IdT, ConstRefType>(IdT(index), value);
+      return std::pair<IdT, ConstRefType>(index, value);
     };
-    return llvm::map_range(llvm::enumerate(values_), index_to_id);
+    return llvm::map_range(llvm::enumerate(values()), index_to_id);
   }
 
  private:
-  // Poison the entire contents of the value store. This is used to detect cases
-  // where references to elements in a value store are used across calls that
-  // might modify the store.
-  auto PoisonAll() const -> void {
-#if CARBON_POISON_VALUE_STORES
-    if (!all_poisoned_) {
-      __asan_poison_memory_region(values_.data(),
-                                  values_.size() * sizeof(values_[0]));
-      all_poisoned_ = true;
-    }
-#endif
-  }
-  // Unpoison the entire contents of the value store.
-  auto UnpoisonAll() const -> void {
-#if CARBON_POISON_VALUE_STORES
-    __asan_unpoison_memory_region(values_.data(),
-                                  values_.size() * sizeof(values_[0]));
-    all_poisoned_ = false;
-#endif
-  }
-  // Poison a single element.
-  auto PoisonElement([[maybe_unused]] int element) const -> void {
-#if CARBON_POISON_VALUE_STORES
-    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
-#endif
-  }
-  // Unpoison a single element.
-  auto UnpoisonElement([[maybe_unused]] int element) const -> void {
-#if CARBON_POISON_VALUE_STORES
-    __asan_unpoison_memory_region(values_.data() + element, sizeof(values_[0]));
-    all_poisoned_ = false;
-#endif
-  }
+  friend class ValueStoreRange<IdT>;
+
+  using ChunkType = Internal::ValueStoreChunk<IdT, ValueType>;
+  // The base chunk, which is always allocated, and where initial values are
+  // added. Once it's full, new values are added into `more_chunks_`. Accessing
+  // values from this chunk only needs to go through a single indirection into
+  // the chunk's array which optimizes the case where there are a small number
+  // of values.
+  ChunkType chunk_;
```

Comment:
Would it be worth indicating the flow in the name, particularly so that the ordering is clear from callers? Note, one thing you could've been doing here is having `chunk_` be the current insertion point -- so I'm really just hoping to disambiguate that. Something like:

```suggestion
  ChunkType first_chunk_;
```
```suggestion
  ChunkType base_chunk_;
```

Reply:
Done